airflow/jobs.py

```diff
@@ -874,177 +882,222 @@ def _execute(self):
         not_ready = set()
         deadlocked = set()
 
-        for task in self.dag.tasks:
-            if (not self.include_adhoc) and task.adhoc:
-                continue
+        # create dag runs
+        dr_start_date = start_date or min([t.start_date for t in self.dag.tasks])
+        next_run_date = self.dag.normalize_schedule(dr_start_date)
+        end_date = end_date or datetime.now()
 
-            start_date = start_date or task.start_date
-            end_date = end_date or task.end_date or datetime.now()
-            for dttm in self.dag.date_range(start_date, end_date=end_date):
-                ti = models.TaskInstance(task, dttm)
-                tasks_to_run[ti.key] = ti
-                session.merge(ti)
-        session.commit()
+        active_dag_runs = []
```

Comment:
The subdag operator passing a flag to the backfill command would be one way to do it. Another would be to make the subdag operator mutate the subdag's max active dag runs. Personally I think subdags should be a first class citizen in airflow (as opposed to being an operator that forks a separate process), but that is for another time.

I think that making the backfill use the scheduler is definitely the place we want to end up. One problem I can see for existing airflow use cases is that it would make it hard to iterate on/test dags since you could not pass a local copy of a dag into a backfill. Once this problem is solved I can't think of any other blockers (at least for AIrbnb).

Note that this end-goal can already be achieved by using the same executor in the airflow.cfg on the backfill machine as the scheduler uses.